app/src/main/java/com/amaze/filemanager/utils/ServiceWatcherUtil.java

```diff
@@ -190,4 +225,22 @@ public void run() {
 
         handler.postDelayed(runnable, 0);
     }
-}
+
+    public interface ServiceWatcherInteractionInterface {
+
+        int STATE_HALTED = 0;
+        int STATE_RESUMED = 1;
+
+        /**
+         * Progress has been halted for some reason
+         */
+        void progressHalted();
+
+        /**
+         * Future extension for possible implementation of pause/resume of services
+         */
+        void progressResumed();
+
+        <T extends Service> T getServiceType();
```

Comment:
Type erasure makes this getter unnecessarily complex, try `Service getService();`, although, seeing this interface is used solely in `ProgressiveService` it seems redundant.

Reply:
Passing service instances within different threads will only lead to complexities, that's why I created a simple interface as the dependency on service was not that major.
Also, the interface is not used in ````ProgressiveService```` but in ````ServiceWatcherUtil````.

Reply:
> Passing service instances within different threads will only lead to complexities, that's why I created a simple interface as the dependency on service was not that major.

When you "pass an interface" you are passing an instance of a class that implements it, in this case the only class that implements it is ProgressiveService, that is a Service, so you are passing a Service around. 

If you want to only pass a certain operations on the service around (and not all of them) you can make an object that contains the Service but only allows certain functions.

But type erasure is still a thing, and `<T extends Service> T getServiceType();` is ridiculous.

> Also, the interface is not used in ProgressiveService but in ServiceWatcherUtil.

By 'used' I meant "only ProgressiveService implements ServiceWatcherUtil".

Reply:
I'm not passing a service instance around The interface is just fine, and does the job it's supposed to do.

Reply:
Ok, pass the object as an instance of the interface, so as to make more difficult unwanted access to other functions.

But the original issue remains: 

> Type erasure makes this [`getServiceType()`] getter unnecessarily complex

In addition `getServiceType()` doesn't actually return a type.